### PR TITLE
Create a reusable search + list control for category selection

### PR DESCRIPTION
**Pull request by ryelle**
This PR introduces two new components, `ProductCategoryControl` and `SearchListControl`. 

`SearchListControl` renders a search box and list of selectable items, along with a list of "selected" items (which are removed from the selectable list) as removable `Tag`s. Removing a selected item puts it back into the main selectable list. Searching using the text box runs a simple text search over the item's `name`. This is a "simple" component, in that it does not do anything with an API or saving attribute data itself.

`ProductCategoryControl` wraps around `SearchListControl` and handles fetching all categories from the API, and also passes through a render function for customizing the item display (which lets us show the product category's count). This also passes through customized `messages`, again letting us specify "category" copy.

The goal with `SearchListControl` is to be reusable, since we'll have a similar interface for handpicked products.

Other additions in the PR:

- Spacing variables from wc-admin
- A new mixin called `visually-hidden`, which applies the same rules as `screen-reader-text` (for when we can't add a class)

I'll add snapshot tests in a following PR, but wanted this up for testing today.

### Screenshots

First, in edit mode, when we start the selected list is empty, so we just see the search box and list of all categories.

edit-mode-empty

After selecting some categories, we see them above, and no longer in the list.

edit-mode-selected

We can search for items and it will highlight the search terms:

edit-mode-search-results

If there are no results, we just say so (this can be improved 🙂 )

edit-mode-no-results

Once we click Done, we see the block preview. Now we can see the same interface in the sidebar

sidebar

And searching here…

sidebar-search

Changes to the selected categories here triggers another API request & updates the block preview.

### Accessibility ⌨️ 

These components are usable via keyboard navigation, all clickable actions are also keyboard-focusable (the search input, selecting items, tag remove icons). Unfortunately this is not working perfectly with a screen reader, when selected items are removed from the list, it pulls focus out of the page 🙁  I'll need to investigate further for any focus management solutions. Given the state of gutenberg + screen readers in general, I think this is OK for this release.

### How to test the changes in this Pull Request:

1. Add a products by category block
2. Try selecting various categories, and searching for others

~⚠️  I haven't tested with 100+ categories, I'm pretty sure this iteration won't work for that many but I'll address that in a following PR.~ Nevermind, `apiFetch` [is smart and automatically paginates!](https://github.com/WordPress/gutenberg/blob/master/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L42)

### CODE REVIEW

**Review by timmyc — COMMENTED** [review]
This is looking quite awesome! Few comments in the code, but great to see some goodies from wc-admin being used over here.

In testing, the one ux/design issue that we might need to consider is how we handle child categories. The Gutenberg Term Selector uses nesting to handle these... in Calypso I think we might have pre-pended the parent category(ies) names, but currently I'm seeing the following:

image

For a product category set that has parent categories for "t-shirts" and "hoodies" and child categories in each for cats and dogs:

image

While a questionable way to categorize products, I have encountered similar tagging approaches in posts in _the wild_. I think we could look at a design for this in a follow-up PR, but figured it would be worth noting now.

File: assets/js/components/product-category-control/index.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		super( ...arguments );
+		this.state = {
+			list: [],
+		};
+		this.renderItem = this.renderItem.bind( this );
+	}
+
+	componentDidMount() {
+		apiFetch( {
+			path: addQueryArgs( '/wc/v3/products/categories', { per_page: -1 } ),
```
[the remark this answers is not in the document]
Comment (ryelle) [reply assets/js/components/product-category-control/index.js]: [In wp-land, `per_page=-1`](https://developer.wordpress.org/reference/classes/wp_query/#pagination-parameters) requests an "unbounded query" (all categories), but apiFetch is smart and [actually automatically paginates requests](https://github.com/WordPress/gutenberg/blob/master/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L42) for us, so we get all categories here but limit requests to 100 per page. (still impressed with that 😆 )

File: assets/js/components/product-category-control/index.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		} )
+			.then( ( list ) => {
+				this.setState( { list } );
+			} )
+			.catch( () => {
+				this.setState( { list: [] } );
```
[the remark this answers is not in the document]
Comment (ryelle) [reply assets/js/components/product-category-control/index.js]: Yeah, empty states in general need some design help.

File: assets/js/components/search-list-control/index.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+	getHighlightedName( name, search ) {
+		if ( ! search ) {
+			return name;
+		}
+		const re = new RegExp( escapeRegExp( search ), 'ig' );
+		return name.replace( re, '<strong>$&</strong>' );
+	}
+
+	fallbackrenderItem( { getHighlightedName, item, onSelect, search } ) {
```
[the remark this answers is not in the document]
Comment (ryelle) [reply assets/js/components/search-list-control/index.js]: Oh, I also mis-capitalized it :/ 

`defaultRenderItem` sounds good to me.

File: assets/js/components/search-list-control/index.js [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	search: PropTypes.string,
+	setState: PropTypes.func,
+	// from withSpokenMessages
+	debouncedSpeak: PropTypes.func,
+	// from withInstanceId
+	instanceId: PropTypes.string,
+};
+
+export default compose( [
+	withState( {
```
[the remark this answers is not in the document]
Comment (ryelle) [reply assets/js/components/search-list-control/index.js]: There are a lot of cool `with*` HoCs in gutenberg – I've used `withState` on some of our example documentation too, much simpler than setting up a full class-style component.

**Review by LevinMedia — DISMISSED** [review]
This is looking fantastic. What a great start!  

I noticed two bugs while testing, I belive they're related: 

1. from the set up screen, I selected three categories. The total number of products was seven. By default the layout is set to three columns and four rows. When the I clicked "Done" and the block was inserted, I should have had two rows of three, and one row with a single product at the bottom. However, there were only two rows of three.  (Six products total) Here's a gif of the bug:

screen recording 2018-11-28 at 07 37 pm

2. Using the sidebar settings, deleted all my categories. Then I changed the layout to display four columns and four rows. Next I picked a category with 13 products. It appeared as expected. I then picked a category with two products. When it was added, only one of the two products showed up in the block. I noted this with a variety of different categories and layout settings.  Here's .gif of the bug:

screen recording 2018-11-28 at 07 18 pm

--------------------------------------------------------------------------------

The search field in the PR diverges a bit from the original design - just curious about your reason for that. 

Design: 

screen shot 2018-11-28 at 7 53 48 pm

PR: 

screen shot 2018-11-28 at 7 47 40 pm

--------------------------------------------------------------------------------

Once the user clicks "done" on the setup screen, can you open the layout setting container by default? 

--------------------------------------------------------------------------------

@timmyc - @jameskoster and I were talking about the child category issue yesterday. Jay, lets plan on sorting that out first thing tomorrow am so we can put it into a follow up PR per Timmy's suggestion.


🎉

**Comment by jameskoster** [comment]
I'm sure it's something in my setup but for some reason I'm not seeing the category list beneath the input. Searching always returns no results. It's like the block can't find my product categories at all.

My product categories are varied. Some are empty, some have descriptions, some have images assigned to them, some are child categories.

I'm running the latest stable WP and latest Gutenberg. All plugins disabled.

--------------------------------------------------------------------------------

Things I noticed;

* The "Clear all" links should be red.
* From the screenshots it looks like the product count is slightly larger than the category name?
* Not sure if you saw, but it'd be nice to include the animations for selecting/deselecting products; animation

--------------------------------------------------------------------------------

Unrelated to this PR but worth mentioning; 

The top/bottom padding of the settings is inconsistent with other blocks;

padding

I see that's because `.components-placeholder` has a min-height. Using `padding: 2em 1em` on our container seems to achieve parity;

padding

We're now aligning the Done button right and including a cancel link so that folks aren't forced to save changes.

done

**Review by timmyc — APPROVED** [review]
Changes are looking good to me. I think we should land this PR in its current state, since it is over 500 lines of code, and a huge portion of the functionality is working. And thinking we just open up issues to track the other items that were not handled in the last round of updates.

**Comment by timmyc** [comment]
> The following things are left to do, but need more design conversation/work in other PRs

One other item for that list is how to handle nested/child category names. Do we show the parents to differentiate duplicate child category names?

image

Also even without duplicate nested names, does having the parent categories shown help ux?

**Comment by ryelle** [comment]
Ok, I'm going to create issues for that list so that we can have each discussion separately, rather than all in this thread :) 

Aside from those things, is this good to merge?

**Pull request merged**